### PR TITLE
Support GHC 9.2

### PR DESCRIPTION
**Pull request by tek**
Caveat: The custom type error for `insertAt` has been disabled for this version since there appears to have been introduced a subtle change in the types used to match `InsertAtIndex` constraint, causing the incoherent instances to break.
Since the custom error feature isn't required for `insertAt` to work, I propose to postpone the fix for this.

Incorporates @spacekitteh's PRs #432 and #433.

### CODE REVIEW

File: polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs [hunk 1]
```diff
@@ -72,7 +71,16 @@ unify solve_ctx = tryUnifyUnivarsButNotSkolems skolems
         InterpreterUse _ s -> s
         FunctionDef s      -> s
 
-
+#if __GLASGOW_HASKELL__ >= 902
+tryUnifyUnivarsButNotSkolems :: Set TyVar -> Type -> Type -> Maybe TCvSubst
+tryUnifyUnivarsButNotSkolems skolems goal inst =
+  case tcUnifyTysFG
+         (bool (const BindMe) (const Apart) . flip S.member skolems)
```
Comment (isovector) [inline polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs:78]: This is the only bit that's different, no? For things like this I like to just define some CPP:

```haskell
#if __GLASGOW_HASKEL__ >= 902
#define BINDME (const BindMe)
#define APART (const Apart)
#else
#define BINDME BindMe
#define APART Skolem
#endif


tryUnifyUnivarsButNotSkolems :: Set TyVar -> Type -> Type -> Maybe TCvSubst
tryUnifyUnivarsButNotSkolems skolems goal inst =
  case tcUnifyTysFG
         (bool BINDME SKOLEM . flip S.member skolems)
         [inst]
         [goal] of
    Unifiable subst -> pure subst
    _               -> Nothing
```

File: polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs [hunk 2]
```diff
@@ -19,9 +19,8 @@ import           GHC.Core.Type
 import           GHC.Core.Unify
 import           GHC.Plugins (Outputable, ppr, parens, text, (<+>))
 #else
-import           Type
-import           Unify
-import           GhcPlugins (Outputable, ppr, parens, text, (<+>))
+import Type (Type, TyVar, eqType, nonDetCmpType, TCvSubst)
+import Unify (tcUnifyTysFG, BindFlag(Skolem, BindMe), UnifyResultM(Unifiable))
```
Comment (isovector) [inline polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs:23]: formatter gone rogue?

**Review by isovector — CHANGES_REQUESTED** [review]
I'd prefer if `tryUnifyUnivarsButNotSkolems` had less duplication between GHC versions. Too easy to get out of sync as it is.

File: polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs (hunk 2 above)
Comment (tek) [reply polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs:23]: also missing an import :grimacing:

File: polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs (hunk 1 above)
Comment (tek) [reply polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs]: :heavy_check_mark:

File: polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs (hunk 2 above)
Comment (tek) [reply polysemy-plugin/src/Polysemy/Plugin/Fundep/Unification.hs]: :heavy_check_mark:

**Review by isovector — APPROVED** [review]
lgtm!

**Comment by spacekitteh** [comment]
So this still works while there isn't a GHC 9.0.2, but it will break then due to the typos.

**Comment by tek** [comment]
if you're referring to the `__GLASGOW_HASKELL__` number, it only contains the first two parts of the version: https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/intro.html#version-numbering

**Comment by spacekitteh** [comment]
...huh. Ok, ignore that then!

**Comment by spacekitteh** [comment]
Oh - polysemy-plugin needs to be pushed to hackage too!

**Comment by isovector** [comment]
oops, yeah. nice catch!

**Comment by isovector** [comment]
done https://github.com/polysemy-research/polysemy/commit/8f469182f04c555b4f9ac1931ee02e3bb2d6921e

**Pull request merged**